pkg/piping/sync_async_piping_manager.go

```diff
@@ -0,0 +1,172 @@
+				}
+			}
+
+		}(ctx, sap.databaseType, asyncNode, rev.String(), val)
+	}
+
+	var wg sync.WaitGroup
+	for _, syncNode := range syncNodes {
+		wg.Add(1)
+		go func(ctx context.Context, databaseType constants.StoreType, name, key, val string) {
+			defer wg.Done()
+			if len(name) < 1 {
+				return
+			}
+			_, rootSpan := otel.Tracer(config.TraceName).Start(ctx, "sync db put")
+			defer rootSpan.End()
+			if database, err := database.FactoryWithNameAndLatency(sap.databaseType, name, 0); err != nil {
+				rootSpan.RecordError(err)
+				rootSpan.SetStatus(codes.Error, err.Error())
+			} else {
+				if _, err := database.Put(key, val); err != nil {
+					rootSpan.RecordError(err)
+					rootSpan.SetStatus(codes.Error, err.Error())
+				}
+			}
+
+		}(ctx, sap.databaseType, syncNode, rev.String(), val)
+	}
+	wg.Wait()
+
+	return nil
+}
+
+func (sap *SyncAsyncPiping) Delete(ctx context.Context, rev index.Revision) error {
+	_, rootSpan := otel.Tracer(config.TraceName).Start(ctx, "SyncAsyncPiping delete")
+	defer rootSpan.End()
+	syncNodes, asyncNodes, err := splitStores(rev.GetNodes())
+	if err != nil {
+		return err
+	}
+
+	for _, asyncNode := range asyncNodes {
+		if len(asyncNode) < 1 {
+			continue
+		}
+		go func(ctx context.Context, databaseType constants.StoreType, name, key string) {
+			_, rootSpan := otel.Tracer(config.TraceName).Start(ctx, "async db delete")
+			defer rootSpan.End()
+			if database, err := database.FactoryWithNameAndLatency(sap.databaseType, name, 0); err != nil {
+				rootSpan.RecordError(err)
+				rootSpan.SetStatus(codes.Error, err.Error())
+			} else {
+				if err := database.Delete(key); err != nil {
+					rootSpan.RecordError(err)
+					rootSpan.SetStatus(codes.Error, err.Error())
+				}
+			}
+
+		}(ctx, sap.databaseType, asyncNode, rev.String())
+	}
+
+	var wg sync.WaitGroup
+	for _, syncNode := range syncNodes {
+		wg.Add(1)
+		go func(ctx context.Context, databaseType constants.StoreType, name, key string) {
+			defer wg.Done()
+			if len(name) < 1 {
+				return
+			}
+			_, rootSpan := otel.Tracer(config.TraceName).Start(ctx, "sync db delete")
+			defer rootSpan.End()
+			if database, err := database.FactoryWithNameAndLatency(sap.databaseType, name, 0); err != nil {
+				rootSpan.RecordError(err)
+				rootSpan.SetStatus(codes.Error, err.Error())
+			} else {
+				if err := database.Delete(key); err != nil {
+					rootSpan.RecordError(err)
+					rootSpan.SetStatus(codes.Error, err.Error())
+				}
+			}
+
+		}(ctx, sap.databaseType, syncNode, rev.String())
+	}
+	wg.Wait()
+
+	return nil
+}
+
+func splitStores(stores []string) ([]string, []string, error) {
+	syncNodes := make([]string, 0)
+	asyncNodes := make([]string, 0)
+	if len(stores) < 1 {
+		return syncNodes, asyncNodes, fmt.Errorf("no stores in the revision")
+	}
+	syncNodes = strings.Split(stores[0], ",")
+	if len(syncNodes) < 1 {
+		return syncNodes, asyncNodes, fmt.Errorf("no sync stores in the revision")
+	}
+	if len(stores) == 2 {
```

Comment:
why 2?

Reply:
the nodes array contains both sync nodes and async nodes. So if the length is 2, which means the rev has async nodes.

Reply:
consider add a comments or turn 2 into a const variable so its meaning is clear.